### PR TITLE
[Placement Group] Add stats to pg scheduling

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR adds instrumentation information for placement group scheduling. Please check `gcs_placement_group_manager.cc` and `gcs.proto` to see what metrics were added.

It also updates `placement_group_table` to return the "live" information rather than stale information in Redis by replying the entry in memory directly. It is needed because although we store the stats in the protobuf, we don't persist them to Redis every time we update them. Since when GCS starts, it loads all alive pgs to memory, this should be safe.

TODO
- [ ] Add unit tests

## Related issue number

https://github.com/ray-project/ray/issues/19098

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:159]: Oh that's right lol. Thanks for catching this

File: src/ray/gcs/gcs_server/gcs_resource_manager.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_resource_manager.cc:88]: Actually, I couldn't find GetCurrentTimeMS() (I could only find Nanos), and it seems like we only use this method in the codebase, so I guess the function is not available? Not 100% sure though

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [reply src/ray/protobuf/gcs.proto:495]: Do you suggest to combine those states or put them together for better readability ?

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:25]: Yes. This is measure after all commit requests are replied, meaning resources are available on all nodes when we record it. I will be clearer in the. comment.

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [reply src/ray/protobuf/gcs.proto:515]: Ah, I used microsecond at first, so I used double to catch microseconds level values. I think since I am using microseconds, int64_t is probably sufficient (creation usually takes about 1 ms it seems like, so catching diff in nano seconds probably has no value.)

File: src/ray/protobuf/gcs.proto
Comment (scv119) [reply src/ray/protobuf/gcs.proto:495]: yup; we need a summary of the state transfer of PG somewhere in the code.

Comment (rkooo567) [reply src/ray/protobuf/gcs.proto:495]: Can you check if the protobuf def makes sense?

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:521]: this looks like a recurring pattern? should we create a helper function for it?

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.h
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.h:145]: 0 looks bad here?

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (scv119) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:566]: pro tip: `for (auto& [pg_id, data] : result)`

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.h
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.h:145]:
```
Bar foo() const: Returns the current value of the field. If the field is not set, returns the default value (0).
```

According to the doc, the default value is 0. I can alternatively just use `rpc::PlacementGroupStats::QUEUED`, but I thought it might be more confusing (like people will wonder why we set this and probably just remove it). I will probably just add a comment instead (that 0 is the default)

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:521]: which part are you talking about? 
```cpp
 auto it = registered_placement_groups_.find(placement_group_id);
if (it != registered_placement_groups_.end()) {
```
=> this one?

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:521]: I just used `registered_placement_groups_ .contains`. Lmk if you meant something else

**Comment by rkooo567** [comment]
cc @scv119

**Review by scv119 — APPROVED** [review]

**Comment by rkooo567** [comment]
Failing test also is failing in the master

**Comment by rkooo567** [comment]
Waiting for approval from gcs.proto owner. cc @wuisawesome

**Review by wuisawesome — APPROVED** [review]

**Pull request merged**
